### PR TITLE
VB: Support protected access for interface members.

### LINKED ISSUE #35827: VB DIM: Unexpected accessibility error while trying to access protected method declared in an interface
```
        <Fact>
        Public Sub MethodImplementation_13()

            Dim csSource =
"
public interface I1
{
    protected static void M1()
    {
        System.Console.WriteLine(""I1.M1"");
    }
}
"
            Dim csCompiation = GetCSharpCompiation(csSource).EmitToImageReference()

            Dim source1 =
<compilation>
    <file name="c.vb"><![CDATA[
Public Class C
    Implements I1
    Shared Sub Main()
        I1.M1()
    End Sub
End Class
]]></file>
</compilation>

            Dim comp1 = CreateCompilation(source1, options:=TestOptions.DebugExe, targetFramework:=TargetFramework.NetStandardLatest, references:={csCompiation})
            CompileAndVerify(comp1, expectedOutput:=If(ExecutionConditionUtil.IsMonoOrCoreClr, "I1.M1", Nothing), verify:=VerifyOnMonoOrCoreClr)
        End Sub
```

Observed:
```
c.vb(4) : error BC30390: 'I1.Sub M1()' is not accessible in this context because it is 'Protected'.
```

Consuming public static methods works

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #35827.

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb [hunk 1]
```diff
@@ -448,36 +400,118 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                    result = True
+                    Exit While
+                End If
+
+                If baseTypeIsInterface Then
+                    GetBaseInterfaces(current, baseInterfaces, interfacesLookedAt, basesBeingResolved)
                 End If
 
-                If basesBeingResolved IsNot Nothing AndAlso basesBeingResolved.Contains(current) Then
+                If basesBeingResolved.InheritsBeingResolvedOpt IsNot Nothing AndAlso basesBeingResolved.InheritsBeingResolvedOpt.Contains(current) Then
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb]: >basesBeingResolved.InheritsBeingResolvedOpt IsNot Nothing [](start = 19, length = 57)

Nit: could just use conditional access.

File: src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb [hunk 2]
```diff
@@ -448,36 +400,118 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
             While current IsNot Nothing
-                If current.Equals(baseType) Then
-                    Return True
+                If baseTypeIsInterface = current.IsInterfaceType() AndAlso current.Equals(baseType) Then
+                    result = True
+                    Exit While
+                End If
+
+                If baseTypeIsInterface Then
+                    GetBaseInterfaces(current, baseInterfaces, interfacesLookedAt, basesBeingResolved)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb]: >GetBaseInterfaces [](start = 20, length = 17)

`Get` here makes me think there should be a return that you're ignoring. Consider something like `FindBaseInterfaces`.

Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb]: Or `PopulateBaseInterfaces`. Just something that doesn't imply a return.

---
In reply to: [315436500](https://github.com/dotnet/roslyn/pull/38103#discussion_r315436500) [](ancestors = 315436500)

**Comment by 333fred** [comment]
Done review pass (commit 1). Just a minor naming comment.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by jcouv** [comment]
Looking

File: src/Compilers/VisualBasic/Portable/Binding/BasesBeingResolvedBinder.vb [hunk 3]
```diff
@@ -20,27 +20,61 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+
+        Public Function PrependInheritsBeingResolved(symbol As TypeSymbol) As BasesBeingResolved
+            Return New BasesBeingResolved(If(InheritsBeingResolvedOpt, ConsList(Of TypeSymbol).Empty).Prepend(symbol), ImplementsBeingResolvedOpt)
+        End Function
+
+        Public Function PrependImplementsBeingResolved(symbol As TypeSymbol) As BasesBeingResolved
+            Return New BasesBeingResolved(InheritsBeingResolvedOpt, If(ImplementsBeingResolvedOpt, ConsList(Of TypeSymbol).Empty).Prepend(symbol))
+        End Function
+
+        Public ReadOnly Property IsEmpty As Boolean
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Binding/BasesBeingResolvedBinder.vb]: >IsEmpty [](start = 33, length = 7)

nit: I'm not sure this property is useful given that the caller still has to check both parts individually and handle `Nothing` case. Maybe we can do without it, or split into two properties (one for inherits and one for implements)?

File: src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb [hunk 4]
```diff
@@ -448,36 +400,118 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                 End If
 
-                If basesBeingResolved IsNot Nothing AndAlso basesBeingResolved.Contains(current) Then
+                If basesBeingResolved.InheritsBeingResolvedOpt?.Contains(current) Then
                     ' We can't obtain BaseType if we're currently resolving the base type of current.
                     current = Nothing ' can't go up the base type chain.
                 Else
+                    Select Case current.TypeKind
+                        Case TypeKind.Interface
+                            current = Nothing
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Semantics/AccessCheck.vb]: >current = Nothing [](start = 28, length = 17)

I didn't understand why we stop here. The interface could inherit other interfaces, one of which might be `baseType`.

File: src/Compilers/VisualBasic/Portable/Binding/BasesBeingResolvedBinder.vb (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Binding/BasesBeingResolvedBinder.vb]: > nit: I'm not sure this property is useful given that the caller still has to check both parts individually and handle Nothing case. Maybe we can do without it, or split into two properties (one for inherits and one for implements)?

It is used only in one place, I'll remove it in another PR.

---
In reply to: [315814272](https://github.com/dotnet/roslyn/pull/38103#discussion_r315814272) [](ancestors = 315814272)

**Pull request merged**